ffi/src/winscard/scard_handle.rs

```diff
@@ -87,26 +92,223 @@ impl WinScardContextHandle {
+        self.write_to_out_buf(&data, buffer_type)
+    }
+
+    /// Saves the provided data in the [OutBuffer] based on the [RequestedBufferType].
+    pub fn write_to_out_buf(
+        &mut self,
+        data: &[u8],
+        buffer_type: RequestedBufferType,
+    ) -> WinScardResult<OutBuffer<'static>> {
+        Ok(match buffer_type {
+            RequestedBufferType::Buff(buf) => {
+                if buf.len() < data.len() {
+                    return Err(
+                        Error::new(
+                            ErrorKind::InsufficientBuffer, format!(
+                                "provided buffer is too small to fill the requested attribute into: buffer len: {}, attribute data len: {}.",
+                                buf.len(),
+                                data.len()
+                            )
+                        )
+                    );
+                }
+
+                buf[0..data.len()].copy_from_slice(data);
+
+                OutBuffer::Written(data.len())
+            }
+            RequestedBufferType::Length => OutBuffer::DataLen(data.len()),
+            RequestedBufferType::Allocate => {
+                let allocated = self.allocate_buffer(data.len())?;
+                // SAFETY: The `allocated` pointer has been returned from the [WinScarfdContextHandle]
+                // internal method, so it's safe to create a slice.
+                let buf = unsafe { from_raw_parts_mut(allocated, data.len()) };
+
+                buf.copy_from_slice(data);
+
+                OutBuffer::Allocated(buf)
+            }
+        })
+    }
 }
 
 impl Drop for WinScardContextHandle {
     fn drop(&mut self) {
         // [SCardReleaseContext](https://learn.microsoft.com/en-us/windows/win32/api/winscard/nf-winscard-scardreleasecontext)
         // ...freeing any resources allocated under that context, including SCARDHANDLE objects
-        unsafe {
-            for scard in &self.scards {
-                let _ = Box::from_raw(*scard as *mut WinScardHandle);
-            }
+        for scard in &self.scards {
+            // SAFETY: The `WinScardContextHandle` contains only valid scard handles, and it should
+            // be safe to cast them to `WinScardHandle` pointer.
+            let _ = unsafe { Box::from_raw(*scard as *mut WinScardHandle) };
         }
         // ...and memory allocated using the SCARD_AUTOALLOCATE length designator.
-        unsafe {
-            for buff in &self.allocations {
+        for buff in &self.allocations {
+            // SAFETY: It's safe to call the `free` function because the `WinScardContextHandle`
+            // contains only allocated memory pointers.
+            unsafe {
                 libc::free(*buff as _);
             }
         }
     }
 }
 
+/// Represents how and what data the user want to extract.
+#[derive(Debug)]
+pub enum RequestedBufferType<'data> {
+    /// This means the user wants the data filled in the provided buffer.
+    Buff(&'data mut [u8]),
```

Comment:
suggestion: The usual abbreviation is "buf", with a single f.
```suggestion
    Buf(&'data mut [u8]),
```

Reply:
Done